EDP-380C-28-R2-Lab3.R

```diff
@@ -12,3 +12,500 @@ source("scripts/rmvnorm.R")
+# b1     2.3076923  2.3128462 -0.0051539037
+# b2    -1.3461538 -1.3470146  0.0008607601
+# SD(e)  4.0191848  4.0154969  0.0036878149
+# R2     0.3538462  0.3542002 -0.0003540325
+
+#-------------------------3.3d------------------------------------------#
+# Generating for Any Number of Predictors
+
+# Create two functions that combine the predictor data generation with the outcome data generation for methods 1 and 2. Both functions should implement the general matrix expressions given so that they can incorporate any number of predictors. These functions should map onto the following inputs:  function(n, p_x, p_y)
+
+# function to create the covariance matrix
+genSigma <- function(params){
+    # create a matrix of zeros with variances on diagonal 
+    diagVariances = matrix(diag(params$sigmas), nrow = length(params$sigmas))
+    # create a matrix with correlations on off diag, 1 on diagonal
+    # start with matrix of correct dims with NAs
+    corMatrix = matrix(NA, nrow = length(params$sigmas), ncol = length(params$sigmas))
+    # put in the diagonal correlations which is always just 1s
+    diag(corMatrix) <- rep(1, length(params$sigmas))
+    # add in the correlations bw variables on the lower triangle
+    corMatrix[lower.tri(corMatrix)] <- params$rhos
+    #and the upper triangle
+    corMatrix[upper.tri(corMatrix)] <- t(corMatrix)[upper.tri(t(corMatrix))]
+    
+    # diagVar * corMatrix * diagVar (eq. 20a -- not labelled directly) = covariance matrix  
+    covarMatrix = diagVariances %*% corMatrix %*% diagVariances    
+    return(covarMatrix)
+}
+
+# create function that takes the x_p and y_p inputs and translates them into a list for the next functions like genSigma
+transFun <- function(x_p, y_p){
+    list(
+        # the y params are first in keeping with the partitioning on pg. 15
+        mus = c(y_p$mus, x_p$mus),
+        sigmas = c(y_p$sigmas, x_p$sigmas),
+        rhos = c(y_p$rhos, x_p$rhos)
+    )
+}
+
+# Method 1, generate from total variance explained
+# Test the Method 1 function using five predictors with n = 100000 that are all correlated 0.15 with variances from 1 to 5.Set all regression coefficients to 1,the R2 =0.5, and μy = 25. Set the seed to 6972.
+
+set.seed(6972)
+n <- 100000
+betas <- c(1,1,1,1,1)
+R2 <- .5
+
+x_p <- list(
+    mus = c(0,0,0,0,0),
+    sigmas = c(sqrt(1), sqrt(2), sqrt(3), sqrt(4), sqrt(5)),
+    rhos = c(.15, .15, .15, .15, .15)
+)
+
+# *for method 1, there is no sigma or correlations given for y but including here to make it general to be used in method 2
+y_p <- list(
+    mus = c(25),
+    sigmas = NULL,
+    rhos = NULL
+)
+
+method1Fun <- function(n, x_p, y_p, betas, R2){
+    #get params into workable form
+    params <- transFun(x_p, y_p)
+    # generate the covariance matrix, Sigma
+    covarMatrix <- genSigma(params)
+    # generate the Xs
+    X <- rmvnorm(n, x_p$mus, covarMatrix)
+    # add intercept
+    X <- cbind(matrix(1, NROW(X)), X)
```

Comment:
Why add an intercept and then remove it below? Just never include it.

Reply:
It's just how the function/code works in this instance. It works fine and I'd rather not change it.